# Titanic competition with TensorFlow Decision Forests

This notebook will take you through the steps needed to train a baseline Gradient Boosted Trees Model using TensorFlow Decision Forests and creating a submission on the Titanic competition. 

This notebook shows:

1. How to do some basic pre-processing. For example, the passenger names will be tokenized, and ticket names will be splitted in parts.
1. How to train a Gradient Boosted Trees (GBT) with default parameters
1. How to train a GBT with improved default parameters
1. How to tune the parameters of a GBTs
1. How to train and ensemble many GBTs

# Imports dependencies

In [1]:
import numpy as np
import pandas as pd
import os

import tensorflow as tf
import tensorflow_decision_forests as tfdf

print(f"Found TF-DF {tfdf.__version__}")

Found TF-DF 1.2.0


# Load dataset

In [2]:
train_df = pd.read_csv("/kaggle/input/titanic/train.csv")
serving_df = pd.read_csv("/kaggle/input/titanic/test.csv")

train_df.head(10)

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,NaN,S
5,6,0,3,"Moran, Mr. James",male,NaN,0,0,330877,8.4583,NaN,Q
6,7,0,1,"McCarthy, Mr. Timothy J",male,54.0,0,0,17463,51.8625,E46,S
7,8,0,3,"Palsson, Master. Gosta Leonard",male,2.0,3,1,349909,21.0750,NaN,S
8,9,1,3,"Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)",female,27.0,0,2,347742,11.1333,NaN,S
9,10,1,2,"Nasser, Mrs. Nicholas (Adele Achem)",female,14.0,1,0,237736,30.0708,NaN,C


# Prepare dataset

We will apply the following transformations on the dataset.

1. Tokenize the names. For example, "Braund, Mr. Owen Harris" will become ["Braund", "Mr.", "Owen", "Harris"].
2. Extract any prefix in the ticket. For example ticket "STON/O2. 3101282" will become "STON/O2." and 3101282.

In [3]:
def preprocess(df):
    df = df.copy()
    
    def normalize_name(x):
        return " ".join([v.strip(",()[].\"'") for v in x.split(" ")])
    
    def ticket_number(x):
        return x.split(" ")[-1]
        
    def ticket_item(x):
        items = x.split(" ")
        if len(items) == 1:
            return "NONE"
        return "_".join(items[0:-1])
    
    df["Name"] = df["Name"].apply(normalize_name)
    df["Ticket_number"] = df["Ticket"].apply(ticket_number)
    df["Ticket_item"] = df["Ticket"].apply(ticket_item)                     
    return df
    
preprocessed_train_df = preprocess(train_df)
preprocessed_serving_df = preprocess(serving_df)

preprocessed_train_df.head(5)

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked,Ticket_number,Ticket_item
0,1,0,3,Braund Mr Owen Harris,male,22.0,1,0,A/5 21171,7.2500,NaN,S,21171,A/5
1,2,1,1,Cumings Mrs John Bradley Florence Briggs Thayer,female,38.0,1,0,PC 17599,71.2833,C85,C,17599,PC
2,3,1,3,Heikkinen Miss Laina,female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S,3101282,STON/O2.
3,4,1,1,Futrelle Mrs Jacques Heath Lily May Peel,female,35.0,1,0,113803,53.1000,C123,S,113803,NONE
4,5,0,3,Allen Mr William Henry,male,35.0,0,0,373450,8.0500,NaN,S,373450,NONE


Let's keep the list of the input features of the model. Notably, we don't want to train our model on the "PassengerId" and "Ticket" features.

In [4]:
input_features = list(preprocessed_train_df.columns)
input_features.remove("Ticket")
input_features.remove("PassengerId")
input_features.remove("Survived")
#input_features.remove("Ticket_number")

print(f"Input features: {input_features}")

Input features: ['Pclass', 'Name', 'Sex', 'Age', 'SibSp', 'Parch', 'Fare', 'Cabin', 'Embarked', 'Ticket_number', 'Ticket_item']


# Convert Pandas dataset to TensorFlow Dataset

In [5]:
def tokenize_names(features, labels=None):
    """Divite the names into tokens. TF-DF can consume text tokens natively."""
    features["Name"] =  tf.strings.split(features["Name"])
    return features, labels

train_ds = tfdf.keras.pd_dataframe_to_tf_dataset(preprocessed_train_df,label="Survived").map(tokenize_names)
serving_ds = tfdf.keras.pd_dataframe_to_tf_dataset(preprocessed_serving_df).map(tokenize_names)

# Train model with default parameters

### Train model

First, we are training a GradientBoostedTreesModel model with the default parameters.

In [6]:
model = tfdf.keras.GradientBoostedTreesModel(
    verbose=0, # Very few logs
    features=[tfdf.keras.FeatureUsage(name=n) for n in input_features],
    exclude_non_specified_features=True, # Only use the features in "features"
    random_seed=1234,
)
model.fit(train_ds)

self_evaluation = model.make_inspector().evaluation()
print(f"Accuracy: {self_evaluation.accuracy} Loss:{self_evaluation.loss}")

[INFO 2026-07-05T04:23:49.189831801+00:00 kernel.cc:1214] Loading model from path /tmp/tmpb7qqngdh/model/ with prefix 0fec8aaac6a24eeb
[INFO 2026-07-05T04:23:49.198740367+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2026-07-05T04:23:49.198802865+00:00 kernel.cc:1046] Use fast generic engine


Please report this to the TensorFlow team. When filing the bug, set the verbosity to 10 (on Linux, `export AUTOGRAPH_VERBOSITY=10`) and attach the full output.
Cause: could not get source code
To silence this warning, decorate the function with @tf.autograph.experimental.do_not_convert
Accuracy: 0.8260869383811951 Loss:0.8608942627906799


# Train model with improved default parameters

Now you'll use some specific parameters when creating the GBT model

In [7]:
model = tfdf.keras.GradientBoostedTreesModel(
    verbose=0, # Very few logs
    features=[tfdf.keras.FeatureUsage(name=n) for n in input_features],
    exclude_non_specified_features=True, # Only use the features in "features"
    
    #num_trees=2000,
    
    # Only for GBT.
    # A bit slower, but great to understand the model.
    # compute_permutation_variable_importance=True,
    
    # Change the default hyper-parameters
    # hyperparameter_template="benchmark_rank1@v1",
    
    #num_trees=1000,
    #tuner=tuner
    
    min_examples=1,
    categorical_algorithm="RANDOM",
    #max_depth=4,
    shrinkage=0.05,
    #num_candidate_attributes_ratio=0.2,
    split_axis="SPARSE_OBLIQUE",
    sparse_oblique_normalization="MIN_MAX",
    sparse_oblique_num_projections_exponent=2.0,
    num_trees=2000,
    #validation_ratio=0.0,
    random_seed=1234,
    
)
model.fit(train_ds)

self_evaluation = model.make_inspector().evaluation()
print(f"Accuracy: {self_evaluation.accuracy} Loss:{self_evaluation.loss}")

[INFO 2026-07-05T04:23:52.638477129+00:00 kernel.cc:1214] Loading model from path /tmp/tmp6obux_w8/model/ with prefix a0d80e42affc43e5
[INFO 2026-07-05T04:23:52.649247865+00:00 decision_forest.cc:661] Model loaded with 33 root(s), 1823 node(s), and 10 input feature(s).
[INFO 2026-07-05T04:23:52.649298183+00:00 kernel.cc:1046] Use fast generic engine


Accuracy: 0.760869562625885 Loss:1.0154211521148682


Let's look at the model and you can also notice the information about variable importance that the model figured out

In [8]:
model.summary()

Model: "gradient_boosted_trees_model_1"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
Total params: 1
Trainable params: 0
Non-trainable params: 1
_________________________________________________________________
Type: "GRADIENT_BOOSTED_TREES"
Task: CLASSIFICATION
Label: "__LABEL"

Input Features (11):
	Age
	Cabin
	Embarked
	Fare
	Name
	Parch
	Pclass
	Sex
	SibSp
	Ticket_item
	Ticket_number

No weights

Variable Importance: INV_MEAN_MIN_DEPTH:
    1.           "Sex"  0.576632 ################
    2.           "Age"  0.364297 #######
    3.          "Fare"  0.278839 ####
    4.          "Name"  0.208548 #
    5. "Ticket_number"  0.180792 
    6.        "Pclass"  0.176962 
    7.         "Parch"  0.176659 
    8.   "Ticket_item"  0.175540 
    9.      "Embarked"  0.172339 
   10.         "SibSp"  0.170442 

Variable Importance: NUM_AS_ROOT:
    1.  "Sex" 28.000000 ################
    2. "Name"  5.000000 

# Make predictions

In [9]:
def prediction_to_kaggle_format(model, threshold=0.5):
    proba_survive = model.predict(serving_ds, verbose=0)[:,0]
    return pd.DataFrame({
        "PassengerId": serving_df["PassengerId"],
        "Survived": (proba_survive >= threshold).astype(int)
    })

def make_submission(kaggle_predictions):
    path="/kaggle/working/submission.csv"
    kaggle_predictions.to_csv(path, index=False)
    print(f"Submission exported to {path}")
    
kaggle_predictions = prediction_to_kaggle_format(model)
make_submission(kaggle_predictions)
!head /kaggle/working/submission.csv

Submission exported to /kaggle/working/submission.csv
PassengerId,Survived
892,0
893,0
894,0
895,0
896,0
897,0
898,0
899,0
900,1


# Training a model with hyperparameter tunning

Hyper-parameter tuning is enabled by specifying the tuner constructor argument of the model. The tuner object contains all the configuration of the tuner (search space, optimizer, trial and objective).


In [10]:
tuner = tfdf.tuner.RandomSearch(num_trials=1000)
tuner.choice("min_examples", [2, 5, 7, 10])
tuner.choice("categorical_algorithm", ["CART", "RANDOM"])

local_search_space = tuner.choice("growing_strategy", ["LOCAL"])
local_search_space.choice("max_depth", [3, 4, 5, 6, 8])

global_search_space = tuner.choice("growing_strategy", ["BEST_FIRST_GLOBAL"], merge=True)
global_search_space.choice("max_num_nodes", [16, 32, 64, 128, 256])

#tuner.choice("use_hessian_gain", [True, False])
tuner.choice("shrinkage", [0.02, 0.05, 0.10, 0.15])
tuner.choice("num_candidate_attributes_ratio", [0.2, 0.5, 0.9, 1.0])


tuner.choice("split_axis", ["AXIS_ALIGNED"])
oblique_space = tuner.choice("split_axis", ["SPARSE_OBLIQUE"], merge=True)
oblique_space.choice("sparse_oblique_normalization",
                     ["NONE", "STANDARD_DEVIATION", "MIN_MAX"])
oblique_space.choice("sparse_oblique_weights", ["BINARY", "CONTINUOUS"])
oblique_space.choice("sparse_oblique_num_projections_exponent", [1.0, 1.5])

# Tune the model. Notice the `tuner=tuner`.
tuned_model = tfdf.keras.GradientBoostedTreesModel(tuner=tuner)
tuned_model.fit(train_ds, verbose=0)

tuned_self_evaluation = tuned_model.make_inspector().evaluation()
print(f"Accuracy: {tuned_self_evaluation.accuracy} Loss:{tuned_self_evaluation.loss}")

Use /tmp/tmpf0ljkm24 as temporary training directory


[INFO 2026-07-05T04:26:12.078325966+00:00 kernel.cc:1214] Loading model from path /tmp/tmpf0ljkm24/model/ with prefix 8af26b4223244727
[INFO 2026-07-05T04:26:12.093122469+00:00 decision_forest.cc:661] Model loaded with 19 root(s), 589 node(s), and 12 input feature(s).
[INFO 2026-07-05T04:26:12.093180056+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesGeneric" built
[INFO 2026-07-05T04:26:12.093210453+00:00 kernel.cc:1046] Use fast generic engine


Accuracy: 0.9178082346916199 Loss:0.6503586769104004


In the last line in the cell above, you can see the accuracy is higher than previously with default parameters and parameters set by hand.

This is the main idea behing hyperparameter tuning.

For more information you can follow this tutorial: [Automated hyper-parameter tuning](https://www.tensorflow.org/decision_forests/tutorials/automatic_tuning_colab)

# Making an ensemble

Here you'll create 100 models with different seeds and combine their results

This approach removes a little bit the random aspects related to creating ML models

In the GBT creation is used the `honest` parameter. It will use different training examples to infer the structure and the leaf values. This regularization technique trades examples for bias estimates.

In [11]:
predictions = None
num_predictions = 0

for i in range(100):
    print(f"i:{i}")
    # Possible models: GradientBoostedTreesModel or RandomForestModel
    model = tfdf.keras.GradientBoostedTreesModel(
        verbose=0, # Very few logs
        features=[tfdf.keras.FeatureUsage(name=n) for n in input_features],
        exclude_non_specified_features=True, # Only use the features in "features"

        #min_examples=1,
        #categorical_algorithm="RANDOM",
        ##max_depth=4,
        #shrinkage=0.05,
        ##num_candidate_attributes_ratio=0.2,
        #split_axis="SPARSE_OBLIQUE",
        #sparse_oblique_normalization="MIN_MAX",
        #sparse_oblique_num_projections_exponent=2.0,
        #num_trees=2000,
        ##validation_ratio=0.0,
        random_seed=i,
        honest=True,
    )
    model.fit(train_ds)
    
    sub_predictions = model.predict(serving_ds, verbose=0)[:,0]
    if predictions is None:
        predictions = sub_predictions
    else:
        predictions += sub_predictions
    num_predictions += 1

predictions/=num_predictions

kaggle_predictions = pd.DataFrame({
        "PassengerId": serving_df["PassengerId"],
        "Survived": (predictions >= 0.5).astype(int)
    })

make_submission(kaggle_predictions)

i:0


[INFO 2026-07-05T04:26:13.094177956+00:00 kernel.cc:1214] Loading model from path /tmp/tmpipjdwlzi/model/ with prefix 90533f7ad7f944a3
[INFO 2026-07-05T04:26:13.099248347+00:00 kernel.cc:1046] Use fast generic engine


i:1


[INFO 2026-07-05T04:26:14.485678724+00:00 kernel.cc:1214] Loading model from path /tmp/tmptmr7ctwh/model/ with prefix 4172db3623154567
[INFO 2026-07-05T04:26:14.507506451+00:00 kernel.cc:1046] Use fast generic engine


i:2


[INFO 2026-07-05T04:26:15.483003333+00:00 kernel.cc:1214] Loading model from path /tmp/tmplrufqww7/model/ with prefix bfff85d703224c42
[INFO 2026-07-05T04:26:15.488232223+00:00 kernel.cc:1046] Use fast generic engine


i:3


[INFO 2026-07-05T04:26:17.056705862+00:00 kernel.cc:1214] Loading model from path /tmp/tmpfrzls73b/model/ with prefix 456815116b5d4758
[INFO 2026-07-05T04:26:17.086485956+00:00 kernel.cc:1046] Use fast generic engine


i:4


[INFO 2026-07-05T04:26:18.097482329+00:00 kernel.cc:1214] Loading model from path /tmp/tmplhy6ir8h/model/ with prefix 309b8b9153564030
[INFO 2026-07-05T04:26:18.103929541+00:00 kernel.cc:1046] Use fast generic engine


i:5


[INFO 2026-07-05T04:26:19.038986835+00:00 kernel.cc:1214] Loading model from path /tmp/tmp9aqz1igj/model/ with prefix 880d11619afd4ad9
[INFO 2026-07-05T04:26:19.042633542+00:00 kernel.cc:1046] Use fast generic engine


i:6


[INFO 2026-07-05T04:26:20.102445035+00:00 kernel.cc:1214] Loading model from path /tmp/tmpswbpu_69/model/ with prefix 70958dbabb274a4f
[INFO 2026-07-05T04:26:20.110959158+00:00 kernel.cc:1046] Use fast generic engine


i:7


[INFO 2026-07-05T04:26:21.545462116+00:00 kernel.cc:1214] Loading model from path /tmp/tmp4_s7mxzt/model/ with prefix 05aac258c54743fb
[INFO 2026-07-05T04:26:21.568238111+00:00 kernel.cc:1046] Use fast generic engine


i:8


[INFO 2026-07-05T04:26:22.702300784+00:00 kernel.cc:1214] Loading model from path /tmp/tmphi_igbcs/model/ with prefix 19c34da71746422a
[INFO 2026-07-05T04:26:22.716147628+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2026-07-05T04:26:22.716207649+00:00 kernel.cc:1046] Use fast generic engine


i:9


[INFO 2026-07-05T04:26:24.358552766+00:00 kernel.cc:1214] Loading model from path /tmp/tmp2klcsxrl/model/ with prefix 7101c89157204404
[INFO 2026-07-05T04:26:24.376645324+00:00 kernel.cc:1046] Use fast generic engine


i:10


[INFO 2026-07-05T04:26:25.406219855+00:00 kernel.cc:1214] Loading model from path /tmp/tmpx0s7kin6/model/ with prefix 7583d75e41de40ce
[INFO 2026-07-05T04:26:25.412946644+00:00 kernel.cc:1046] Use fast generic engine


i:11


[INFO 2026-07-05T04:26:26.662824042+00:00 kernel.cc:1214] Loading model from path /tmp/tmpildgolu_/model/ with prefix c371af7129914eed
[INFO 2026-07-05T04:26:26.679453994+00:00 kernel.cc:1046] Use fast generic engine


i:12


[INFO 2026-07-05T04:26:27.686907768+00:00 kernel.cc:1214] Loading model from path /tmp/tmpjdtf1ty7/model/ with prefix 0d8e0e8fd5e24a93
[INFO 2026-07-05T04:26:27.693819386+00:00 kernel.cc:1046] Use fast generic engine


i:13


[INFO 2026-07-05T04:26:28.844190933+00:00 kernel.cc:1214] Loading model from path /tmp/tmpbz0llnns/model/ with prefix 177d90bb701d4bd4
[INFO 2026-07-05T04:26:28.856838399+00:00 kernel.cc:1046] Use fast generic engine


i:14


[INFO 2026-07-05T04:26:29.865143052+00:00 kernel.cc:1214] Loading model from path /tmp/tmp3_cdzbg9/model/ with prefix e77884f064944b33
[INFO 2026-07-05T04:26:29.872074606+00:00 kernel.cc:1046] Use fast generic engine


i:15


[INFO 2026-07-05T04:26:30.9225629+00:00 kernel.cc:1214] Loading model from path /tmp/tmpdpcuuqn2/model/ with prefix f51276fc138b4bbb
[INFO 2026-07-05T04:26:30.931231539+00:00 kernel.cc:1046] Use fast generic engine


i:16


[INFO 2026-07-05T04:26:32.139297665+00:00 kernel.cc:1214] Loading model from path /tmp/tmpvvo6dski/model/ with prefix 9c3e97c29e624b3a
[INFO 2026-07-05T04:26:32.153831973+00:00 kernel.cc:1046] Use fast generic engine


i:17


[INFO 2026-07-05T04:26:33.404820253+00:00 kernel.cc:1214] Loading model from path /tmp/tmptd8jaj8b/model/ with prefix 3c126ccc6c354d56
[INFO 2026-07-05T04:26:33.420104881+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2026-07-05T04:26:33.420168307+00:00 kernel.cc:1046] Use fast generic engine


i:18


[INFO 2026-07-05T04:26:34.599195492+00:00 kernel.cc:1214] Loading model from path /tmp/tmptqf87_s6/model/ with prefix 75614da0a4c8435c
[INFO 2026-07-05T04:26:34.612707081+00:00 kernel.cc:1046] Use fast generic engine


i:19


[INFO 2026-07-05T04:26:35.977983448+00:00 kernel.cc:1214] Loading model from path /tmp/tmpk165tm8w/model/ with prefix 8e9d63fab132440a
[INFO 2026-07-05T04:26:35.998757677+00:00 kernel.cc:1046] Use fast generic engine


i:20


[INFO 2026-07-05T04:26:37.285497753+00:00 kernel.cc:1214] Loading model from path /tmp/tmp7haeg6he/model/ with prefix 8c58aeb8bd3b4352
[INFO 2026-07-05T04:26:37.302768731+00:00 kernel.cc:1046] Use fast generic engine


i:21


[INFO 2026-07-05T04:26:38.295323149+00:00 kernel.cc:1214] Loading model from path /tmp/tmpgha0_uql/model/ with prefix c72a3d44614745c2
[INFO 2026-07-05T04:26:38.301446678+00:00 kernel.cc:1046] Use fast generic engine


i:22


[INFO 2026-07-05T04:26:39.300024761+00:00 kernel.cc:1214] Loading model from path /tmp/tmpe80sq6f1/model/ with prefix 4d36d441a08244d2
[INFO 2026-07-05T04:26:39.306696746+00:00 kernel.cc:1046] Use fast generic engine


i:23


[INFO 2026-07-05T04:26:40.396044826+00:00 kernel.cc:1214] Loading model from path /tmp/tmp8jz22asw/model/ with prefix a9f07492047d485a
[INFO 2026-07-05T04:26:40.406188819+00:00 kernel.cc:1046] Use fast generic engine


i:24


[INFO 2026-07-05T04:26:41.42143445+00:00 kernel.cc:1214] Loading model from path /tmp/tmpstztx6mm/model/ with prefix 7fe91e491f514ea7
[INFO 2026-07-05T04:26:41.428202865+00:00 kernel.cc:1046] Use fast generic engine


i:25


[INFO 2026-07-05T04:26:42.626109626+00:00 kernel.cc:1214] Loading model from path /tmp/tmpa6nqz17k/model/ with prefix 8b127bc98c464ee9
[INFO 2026-07-05T04:26:42.639301674+00:00 kernel.cc:1046] Use fast generic engine


i:26


[INFO 2026-07-05T04:26:43.794344137+00:00 kernel.cc:1214] Loading model from path /tmp/tmpikpaio1b/model/ with prefix 7d77bc520c1d4229
[INFO 2026-07-05T04:26:43.805679904+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2026-07-05T04:26:43.805724016+00:00 kernel.cc:1046] Use fast generic engine


i:27


[INFO 2026-07-05T04:26:44.836085622+00:00 kernel.cc:1214] Loading model from path /tmp/tmp62ogjwjv/model/ with prefix 315c7c8bebff48dd
[INFO 2026-07-05T04:26:44.843629483+00:00 kernel.cc:1046] Use fast generic engine


i:28


[INFO 2026-07-05T04:26:45.838463501+00:00 kernel.cc:1214] Loading model from path /tmp/tmpm1j7vp1c/model/ with prefix ae46815f420644d7
[INFO 2026-07-05T04:26:45.844303977+00:00 kernel.cc:1046] Use fast generic engine


i:29


[INFO 2026-07-05T04:26:47.058699435+00:00 kernel.cc:1214] Loading model from path /tmp/tmpe3bd75ld/model/ with prefix d98351418c424d59
[INFO 2026-07-05T04:26:47.073486745+00:00 kernel.cc:1046] Use fast generic engine


i:30


[INFO 2026-07-05T04:26:48.705384148+00:00 kernel.cc:1214] Loading model from path /tmp/tmpeb49qha2/model/ with prefix 1f448f54aabe4322
[INFO 2026-07-05T04:26:48.7363308+00:00 kernel.cc:1046] Use fast generic engine


i:31


[INFO 2026-07-05T04:26:50.289062482+00:00 kernel.cc:1214] Loading model from path /tmp/tmpha7dq_to/model/ with prefix 00dfc4f265ea4027
[INFO 2026-07-05T04:26:50.301766901+00:00 kernel.cc:1046] Use fast generic engine


i:32


[INFO 2026-07-05T04:26:51.347903481+00:00 kernel.cc:1214] Loading model from path /tmp/tmpvcklyit9/model/ with prefix ba4d577e195e4459
[INFO 2026-07-05T04:26:51.355000795+00:00 kernel.cc:1046] Use fast generic engine


i:33


[INFO 2026-07-05T04:26:52.601886004+00:00 kernel.cc:1214] Loading model from path /tmp/tmpwslx6244/model/ with prefix 93f265daed824715
[INFO 2026-07-05T04:26:52.617261851+00:00 kernel.cc:1046] Use fast generic engine


i:34


[INFO 2026-07-05T04:26:53.733190977+00:00 kernel.cc:1214] Loading model from path /tmp/tmpjc5il129/model/ with prefix 9d667046f78b4c9c
[INFO 2026-07-05T04:26:53.743159094+00:00 kernel.cc:1046] Use fast generic engine


i:35


[INFO 2026-07-05T04:26:54.815017765+00:00 kernel.cc:1214] Loading model from path /tmp/tmphjggs0jo/model/ with prefix f4ba7a5466ad4d0e
[INFO 2026-07-05T04:26:54.824236166+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2026-07-05T04:26:54.824292226+00:00 kernel.cc:1046] Use fast generic engine


i:36


[INFO 2026-07-05T04:26:56.075206073+00:00 kernel.cc:1214] Loading model from path /tmp/tmpw5_fondp/model/ with prefix c9b3254a26324bb8
[INFO 2026-07-05T04:26:56.09146604+00:00 kernel.cc:1046] Use fast generic engine


i:37


[INFO 2026-07-05T04:26:57.190579163+00:00 kernel.cc:1214] Loading model from path /tmp/tmpp1ryrfjz/model/ with prefix 67b1959fd37446ec
[INFO 2026-07-05T04:26:57.200702012+00:00 kernel.cc:1046] Use fast generic engine


i:38


[INFO 2026-07-05T04:26:58.425076447+00:00 kernel.cc:1214] Loading model from path /tmp/tmpjtcza31t/model/ with prefix c7a82ef76672473b
[INFO 2026-07-05T04:26:58.440776444+00:00 kernel.cc:1046] Use fast generic engine


i:39


[INFO 2026-07-05T04:26:59.658193098+00:00 kernel.cc:1214] Loading model from path /tmp/tmpilsikruu/model/ with prefix 027209010c8d4662
[INFO 2026-07-05T04:26:59.67285497+00:00 kernel.cc:1046] Use fast generic engine


i:40


[INFO 2026-07-05T04:27:00.642006392+00:00 kernel.cc:1214] Loading model from path /tmp/tmp0e_36b09/model/ with prefix 57604889fc4f4bc1
[INFO 2026-07-05T04:27:00.647185507+00:00 kernel.cc:1046] Use fast generic engine


i:41


[INFO 2026-07-05T04:27:01.924378178+00:00 kernel.cc:1214] Loading model from path /tmp/tmpmpsizp0j/model/ with prefix 84a2b4f6c12143a2
[INFO 2026-07-05T04:27:01.942440901+00:00 kernel.cc:1046] Use fast generic engine


i:42


[INFO 2026-07-05T04:27:03.095548403+00:00 kernel.cc:1214] Loading model from path /tmp/tmpmohg07zg/model/ with prefix c218231d5b034cbe
[INFO 2026-07-05T04:27:03.106150171+00:00 kernel.cc:1046] Use fast generic engine


i:43


[INFO 2026-07-05T04:27:04.448797796+00:00 kernel.cc:1214] Loading model from path /tmp/tmplgqm_ikt/model/ with prefix 5c861c8fbcfc4e73
[INFO 2026-07-05T04:27:04.468626055+00:00 kernel.cc:1046] Use fast generic engine


i:44


[INFO 2026-07-05T04:27:05.583903727+00:00 kernel.cc:1214] Loading model from path /tmp/tmpatf8pce3/model/ with prefix 66b7a1582ab54111
[INFO 2026-07-05T04:27:05.594798781+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2026-07-05T04:27:05.594853874+00:00 kernel.cc:1046] Use fast generic engine


i:45


[INFO 2026-07-05T04:27:06.546767512+00:00 kernel.cc:1214] Loading model from path /tmp/tmpbnwthl4w/model/ with prefix 50baaf33ed1747a1
[INFO 2026-07-05T04:27:06.550772572+00:00 kernel.cc:1046] Use fast generic engine


i:46


[INFO 2026-07-05T04:27:07.823080556+00:00 kernel.cc:1214] Loading model from path /tmp/tmpue64nvvi/model/ with prefix ce376dbffaae4f63
[INFO 2026-07-05T04:27:07.839810072+00:00 kernel.cc:1046] Use fast generic engine


i:47


[INFO 2026-07-05T04:27:09.079696733+00:00 kernel.cc:1214] Loading model from path /tmp/tmp4z9s75o0/model/ with prefix bf903fd2435e458a
[INFO 2026-07-05T04:27:09.094772975+00:00 kernel.cc:1046] Use fast generic engine


i:48


[INFO 2026-07-05T04:27:10.089042082+00:00 kernel.cc:1214] Loading model from path /tmp/tmp3cxij3r5/model/ with prefix f57824648499470a
[INFO 2026-07-05T04:27:10.094278792+00:00 kernel.cc:1046] Use fast generic engine


i:49


[INFO 2026-07-05T04:27:11.142068262+00:00 kernel.cc:1214] Loading model from path /tmp/tmp6rglkp6f/model/ with prefix 81dec5bcecd94123
[INFO 2026-07-05T04:27:11.149933214+00:00 kernel.cc:1046] Use fast generic engine


i:50


[INFO 2026-07-05T04:27:12.347540551+00:00 kernel.cc:1214] Loading model from path /tmp/tmpgfzl89_1/model/ with prefix 00a1413944394aba
[INFO 2026-07-05T04:27:12.361610807+00:00 kernel.cc:1046] Use fast generic engine


i:51


[INFO 2026-07-05T04:27:13.695055365+00:00 kernel.cc:1214] Loading model from path /tmp/tmpjv5foi1q/model/ with prefix 6bf075dd0e76486f
[INFO 2026-07-05T04:27:13.713980327+00:00 kernel.cc:1046] Use fast generic engine


i:52


[INFO 2026-07-05T04:27:14.800830015+00:00 kernel.cc:1214] Loading model from path /tmp/tmproji5xfy/model/ with prefix deb97a7053f14f43
[INFO 2026-07-05T04:27:14.80964468+00:00 kernel.cc:1046] Use fast generic engine


i:53


[INFO 2026-07-05T04:27:15.87840596+00:00 kernel.cc:1214] Loading model from path /tmp/tmphc0wb8n6/model/ with prefix d802c3b8cca64422
[INFO 2026-07-05T04:27:15.887031095+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2026-07-05T04:27:15.887080681+00:00 kernel.cc:1046] Use fast generic engine


i:54


[INFO 2026-07-05T04:27:16.845179381+00:00 kernel.cc:1214] Loading model from path /tmp/tmpkoaicsi4/model/ with prefix c6ba956be3214b1f
[INFO 2026-07-05T04:27:16.848969033+00:00 kernel.cc:1046] Use fast generic engine


i:55


[INFO 2026-07-05T04:27:18.123795516+00:00 kernel.cc:1214] Loading model from path /tmp/tmp80h4_zq9/model/ with prefix 95a85dbf285b47d9
[INFO 2026-07-05T04:27:18.140738007+00:00 kernel.cc:1046] Use fast generic engine


i:56


[INFO 2026-07-05T04:27:19.778031184+00:00 kernel.cc:1214] Loading model from path /tmp/tmppnirc1ni/model/ with prefix 8771c51f89df41bb
[INFO 2026-07-05T04:27:19.792593058+00:00 kernel.cc:1046] Use fast generic engine


i:57


[INFO 2026-07-05T04:27:20.818256117+00:00 kernel.cc:1214] Loading model from path /tmp/tmpghxgsudv/model/ with prefix da72ac56b1d84b3d
[INFO 2026-07-05T04:27:20.82333607+00:00 kernel.cc:1046] Use fast generic engine


i:58


[INFO 2026-07-05T04:27:21.880892161+00:00 kernel.cc:1214] Loading model from path /tmp/tmp_lj56iad/model/ with prefix b08dea3895e54093
[INFO 2026-07-05T04:27:21.88863027+00:00 kernel.cc:1046] Use fast generic engine


i:59


[INFO 2026-07-05T04:27:23.046116098+00:00 kernel.cc:1214] Loading model from path /tmp/tmpi2ttylzw/model/ with prefix 532e4bac19cd4089
[INFO 2026-07-05T04:27:23.057095352+00:00 kernel.cc:1046] Use fast generic engine


i:60


[INFO 2026-07-05T04:27:24.205415643+00:00 kernel.cc:1214] Loading model from path /tmp/tmpzh29o04i/model/ with prefix 1558b7463f014c2f
[INFO 2026-07-05T04:27:24.216756147+00:00 kernel.cc:1046] Use fast generic engine


i:61


[INFO 2026-07-05T04:27:25.227073803+00:00 kernel.cc:1214] Loading model from path /tmp/tmpeay8wrvm/model/ with prefix 4dbc088a30c94bc0
[INFO 2026-07-05T04:27:25.232785796+00:00 kernel.cc:1046] Use fast generic engine


i:62


[INFO 2026-07-05T04:27:26.820586715+00:00 kernel.cc:1214] Loading model from path /tmp/tmpgu_jcn8d/model/ with prefix 1530ab5941c0431d
[INFO 2026-07-05T04:27:26.850233223+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2026-07-05T04:27:26.85029303+00:00 kernel.cc:1046] Use fast generic engine


i:63


[INFO 2026-07-05T04:27:27.962199279+00:00 kernel.cc:1214] Loading model from path /tmp/tmpkzca62_3/model/ with prefix 0b710ffe667b4fab
[INFO 2026-07-05T04:27:27.972826945+00:00 kernel.cc:1046] Use fast generic engine


i:64


[INFO 2026-07-05T04:27:29.051276614+00:00 kernel.cc:1214] Loading model from path /tmp/tmpoc59w_93/model/ with prefix f9015082248f413e
[INFO 2026-07-05T04:27:29.060424621+00:00 kernel.cc:1046] Use fast generic engine


i:65


[INFO 2026-07-05T04:27:30.052473377+00:00 kernel.cc:1214] Loading model from path /tmp/tmp2dlb2jg9/model/ with prefix 7e6861a9de90426e
[INFO 2026-07-05T04:27:30.058081988+00:00 kernel.cc:1046] Use fast generic engine


i:66


[INFO 2026-07-05T04:27:31.101045812+00:00 kernel.cc:1214] Loading model from path /tmp/tmp07vmpgrr/model/ with prefix c7fb7d355ef648a6
[INFO 2026-07-05T04:27:31.108370884+00:00 kernel.cc:1046] Use fast generic engine


i:67


[INFO 2026-07-05T04:27:32.488242264+00:00 kernel.cc:1214] Loading model from path /tmp/tmplyiar6gx/model/ with prefix 9f565dde8f2240f7
[INFO 2026-07-05T04:27:32.510735853+00:00 kernel.cc:1046] Use fast generic engine


i:68


[INFO 2026-07-05T04:27:33.726004877+00:00 kernel.cc:1214] Loading model from path /tmp/tmpv1tfgk7t/model/ with prefix a5cbd5dd91724fe6
[INFO 2026-07-05T04:27:33.739483694+00:00 kernel.cc:1046] Use fast generic engine


i:69


[INFO 2026-07-05T04:27:34.788426234+00:00 kernel.cc:1214] Loading model from path /tmp/tmpvhm9gew7/model/ with prefix 18136dac263848b7
[INFO 2026-07-05T04:27:34.795896054+00:00 kernel.cc:1046] Use fast generic engine


i:70


[INFO 2026-07-05T04:27:35.901336839+00:00 kernel.cc:1214] Loading model from path /tmp/tmpz1reoa4z/model/ with prefix 0cf31127fdc44e6e
[INFO 2026-07-05T04:27:35.911257464+00:00 kernel.cc:1046] Use fast generic engine


i:71


[INFO 2026-07-05T04:27:36.996442895+00:00 kernel.cc:1214] Loading model from path /tmp/tmp79kcj407/model/ with prefix 054e706d79db45b9
[INFO 2026-07-05T04:27:37.004962158+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2026-07-05T04:27:37.005004045+00:00 kernel.cc:1046] Use fast generic engine


i:72


[INFO 2026-07-05T04:27:38.318923578+00:00 kernel.cc:1214] Loading model from path /tmp/tmphdhumk8w/model/ with prefix a4c1dffbb2334b35
[INFO 2026-07-05T04:27:38.336870302+00:00 kernel.cc:1046] Use fast generic engine


i:73


[INFO 2026-07-05T04:27:39.408187514+00:00 kernel.cc:1214] Loading model from path /tmp/tmpdi60ncgk/model/ with prefix 0e340a3332594757
[INFO 2026-07-05T04:27:39.416166243+00:00 kernel.cc:1046] Use fast generic engine


i:74


[INFO 2026-07-05T04:27:40.611454378+00:00 kernel.cc:1214] Loading model from path /tmp/tmpyl_nspp5/model/ with prefix 4431e11705114c77
[INFO 2026-07-05T04:27:40.624723531+00:00 kernel.cc:1046] Use fast generic engine


i:75


[INFO 2026-07-05T04:27:41.719364971+00:00 kernel.cc:1214] Loading model from path /tmp/tmpozj6zrst/model/ with prefix e6c17cc26d984730
[INFO 2026-07-05T04:27:41.728636636+00:00 kernel.cc:1046] Use fast generic engine


i:76


[INFO 2026-07-05T04:27:42.72159896+00:00 kernel.cc:1214] Loading model from path /tmp/tmpfyyrro2w/model/ with prefix df207e544d9e4590
[INFO 2026-07-05T04:27:42.726708166+00:00 kernel.cc:1046] Use fast generic engine


i:77


[INFO 2026-07-05T04:27:43.742481871+00:00 kernel.cc:1214] Loading model from path /tmp/tmpn47ppvkk/model/ with prefix e79a9f414e494aef
[INFO 2026-07-05T04:27:43.747842287+00:00 kernel.cc:1046] Use fast generic engine


i:78


[INFO 2026-07-05T04:27:44.802889136+00:00 kernel.cc:1214] Loading model from path /tmp/tmpwrmcjalz/model/ with prefix d176e744f89f4d97
[INFO 2026-07-05T04:27:44.810489337+00:00 kernel.cc:1046] Use fast generic engine


i:79


[INFO 2026-07-05T04:27:45.86695549+00:00 kernel.cc:1214] Loading model from path /tmp/tmpl9oavvb_/model/ with prefix 88e5a50104204449
[INFO 2026-07-05T04:27:45.875012147+00:00 kernel.cc:1046] Use fast generic engine


i:80


[INFO 2026-07-05T04:27:47.01709896+00:00 kernel.cc:1214] Loading model from path /tmp/tmp5erd48p2/model/ with prefix eb01bc898222468d
[INFO 2026-07-05T04:27:47.028393024+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2026-07-05T04:27:47.028438812+00:00 kernel.cc:1046] Use fast generic engine


i:81


[INFO 2026-07-05T04:27:48.217083553+00:00 kernel.cc:1214] Loading model from path /tmp/tmplrwoer0p/model/ with prefix ae7e758062c04b16
[INFO 2026-07-05T04:27:48.22991446+00:00 kernel.cc:1046] Use fast generic engine


i:82


[INFO 2026-07-05T04:27:49.36247084+00:00 kernel.cc:1214] Loading model from path /tmp/tmpcq8p0krr/model/ with prefix 9299f02ef19f410a
[INFO 2026-07-05T04:27:49.373677653+00:00 kernel.cc:1046] Use fast generic engine


i:83


[INFO 2026-07-05T04:27:50.487773475+00:00 kernel.cc:1214] Loading model from path /tmp/tmpvxhm_b31/model/ with prefix 7829e03155ee44c8
[INFO 2026-07-05T04:27:50.497921067+00:00 kernel.cc:1046] Use fast generic engine


i:84


[INFO 2026-07-05T04:27:52.380306154+00:00 kernel.cc:1214] Loading model from path /tmp/tmposcnj0ux/model/ with prefix 5c7ce8a130234e48
[INFO 2026-07-05T04:27:52.402147594+00:00 kernel.cc:1046] Use fast generic engine


i:85


[INFO 2026-07-05T04:27:53.514873435+00:00 kernel.cc:1214] Loading model from path /tmp/tmphu96dzwj/model/ with prefix 815938c69b614178
[INFO 2026-07-05T04:27:53.52278905+00:00 kernel.cc:1046] Use fast generic engine


i:86


[INFO 2026-07-05T04:27:54.85443308+00:00 kernel.cc:1214] Loading model from path /tmp/tmpcs2v7d3p/model/ with prefix 567e58c2341f4e2b
[INFO 2026-07-05T04:27:54.873281081+00:00 kernel.cc:1046] Use fast generic engine


i:87


[INFO 2026-07-05T04:27:56.256684064+00:00 kernel.cc:1214] Loading model from path /tmp/tmptzmmudi1/model/ with prefix 9c719bc89c1f45b8
[INFO 2026-07-05T04:27:56.277272239+00:00 kernel.cc:1046] Use fast generic engine


i:88


[INFO 2026-07-05T04:27:57.466210152+00:00 kernel.cc:1214] Loading model from path /tmp/tmprm1pijn_/model/ with prefix a240930e6322405e
[INFO 2026-07-05T04:27:57.479551548+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2026-07-05T04:27:57.479609482+00:00 kernel.cc:1046] Use fast generic engine


i:89


[INFO 2026-07-05T04:27:58.491074735+00:00 kernel.cc:1214] Loading model from path /tmp/tmpd59c2s4l/model/ with prefix 3a9064d0de03479e
[INFO 2026-07-05T04:27:58.496815394+00:00 kernel.cc:1046] Use fast generic engine


i:90


[INFO 2026-07-05T04:27:59.646805898+00:00 kernel.cc:1214] Loading model from path /tmp/tmp98q5r7c0/model/ with prefix aa01529ea7bc46c4
[INFO 2026-07-05T04:27:59.658607744+00:00 kernel.cc:1046] Use fast generic engine


i:91


[INFO 2026-07-05T04:28:00.709487408+00:00 kernel.cc:1214] Loading model from path /tmp/tmpegqsr88j/model/ with prefix c41d1a9b8c544d92
[INFO 2026-07-05T04:28:00.717315274+00:00 kernel.cc:1046] Use fast generic engine


i:92


[INFO 2026-07-05T04:28:02.072556419+00:00 kernel.cc:1214] Loading model from path /tmp/tmpzpcyoupf/model/ with prefix 91ee733d10d64d55
[INFO 2026-07-05T04:28:02.092699997+00:00 kernel.cc:1046] Use fast generic engine


i:93


[INFO 2026-07-05T04:28:03.273359696+00:00 kernel.cc:1214] Loading model from path /tmp/tmpbpm1eij_/model/ with prefix b2807b779a64445b
[INFO 2026-07-05T04:28:03.285150188+00:00 kernel.cc:1046] Use fast generic engine


i:94


[INFO 2026-07-05T04:28:04.326689903+00:00 kernel.cc:1214] Loading model from path /tmp/tmp2efw36v6/model/ with prefix 88a8788d3ef543d1
[INFO 2026-07-05T04:28:04.334162951+00:00 kernel.cc:1046] Use fast generic engine


i:95


[INFO 2026-07-05T04:28:05.437877659+00:00 kernel.cc:1214] Loading model from path /tmp/tmphtp9eyhj/model/ with prefix 8a140e70a71d4ab7
[INFO 2026-07-05T04:28:05.447473516+00:00 kernel.cc:1046] Use fast generic engine


i:96


[INFO 2026-07-05T04:28:06.597605819+00:00 kernel.cc:1214] Loading model from path /tmp/tmp82fdgr0v/model/ with prefix 5abc47b500ca4047
[INFO 2026-07-05T04:28:06.609116733+00:00 kernel.cc:1046] Use fast generic engine


i:97


[INFO 2026-07-05T04:28:07.630682215+00:00 kernel.cc:1214] Loading model from path /tmp/tmpkh2ygb59/model/ with prefix cde63b9732264700
[INFO 2026-07-05T04:28:07.636605505+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2026-07-05T04:28:07.63664718+00:00 kernel.cc:1046] Use fast generic engine


i:98


[INFO 2026-07-05T04:28:08.718570025+00:00 kernel.cc:1214] Loading model from path /tmp/tmp9sys5pys/model/ with prefix d5a77817805a429c
[INFO 2026-07-05T04:28:08.727660134+00:00 kernel.cc:1046] Use fast generic engine


i:99


[INFO 2026-07-05T04:28:10.007306859+00:00 kernel.cc:1214] Loading model from path /tmp/tmpmj370sh6/model/ with prefix 608b8ef86f9347a0
[INFO 2026-07-05T04:28:10.023361552+00:00 kernel.cc:1046] Use fast generic engine


Submission exported to /kaggle/working/submission.csv
